blog/_posts/2019-01-18-fluxdiffeq.md

```diff
@@ -0,0 +1,650 @@
+it puts it in the differential equation defined by `prob`, solves it with
+the chosen arguments (solver, tolerance, etc), and then gets a vector out through
+some function `reduction`. For example, in our Lotka-Volerra problem, let's say we wanted
+to go from `p` to a vector of an evenly-spaced time series for `x(t)`.
+In the REPL, we would write:
+
+```julia
+p = [1.5,1.0,3.0,1.0]
+prob = ODEProblem(lotka_volterra,u0,tspan,p)
+sol = solve(prob,Tsit5(),saveat=0.1)
+A = sol[1,:] # length 101 vector
+```
+
+Let's plot `(t,A)` over the ODE's solution to see what we got:
+
+```julia
+plot(sol)
+t = 0:0.1:10.0
+scatter!(t,A)
+```
+
+![Data points plot](https://user-images.githubusercontent.com/1814174/51388173-9c6a4d00-1af6-11e9-9878-3c585d3cfffe.png)
+
+So this is what we wanted: a vector which is the time series of `x(t)` saved at every 0.1 time steps
+given that the parameters are `p`. To write this using the layer function, we simply make
+our solution handling be in the function `reduction` used above in `diffeq_rd(p,reduction,prob,args...;kwargs...)`.
+Thus this code is equivalent to:
+
+```julia
+using FluxDiffEq
+reduction(sol) = sol[1,:]
+diffeq_rd(p,reduction,prob,Tsit5(),saveat=0.1)
+```
+
+As a reminder, this is a function which takes in a parameter vector `p`,
+solves the Lotka-Volterra problem `prob` with our chosen ODE solver arguments,
+and then gets a vector out via `reduction` which is the time series of `x(t)`.
+
+The nice thing about `diffeq_rd` is that it takes care of the type handling
+necessary to make it compatible with the neural network framework (here Flux.jl). To show this,
+let's define a neural network with the function as our single layer, and then a loss
+function that is the squared distance of the output values from `1`. In Flux, this looks like:
+
+```julia
+p = param([2.2, 1.0, 2.0, 0.4]) # Initial Parameter Vector
+params = Flux.Params([p])
+
+function predict_rd() # Our 1-layer neural network
+  diffeq_rd(p,reduction,prob,Tsit5(),saveat=0.1)
+end
+
+loss_rd() = sum(abs2,x-1 for x in predict_rd()) # loss function
+```
+
+Now we tell Flux to train the neural network by running a 100 epoch
+to minimise our loss function (`loss_rd()`) and thus obtain the optimized parameters:
+
+```julia
+data = Iterators.repeated((), 100)
+opt = ADAM(0.1)
+cb = function () #callback function to observe training
+  display(loss_rd())
+  # using `remake` to re-create our `prob` with current parameters `p`
+  display(plot(solve(remake(prob,p=Flux.data(p)),Tsit5(),saveat=0.1),ylim=(0,6)))
+end
+
+# Display the ODE with the initial parameter values.
+cb()
+
+Flux.train!(loss_rd, params, data, opt, cb = cb)
+```
+
+The result of this is the animation shown at the top.
+
+Flux.jl finds the parameters of the neural network (`p`) which minimize
+the cost function, i.e. it trains the neural network: it just so happens that here training the neural network happens to include solving an ODE.

```

Comment:
"training the neural network happens to include solving an ODE" kinda implies that it's the optimizer that requires solving the ODE, not the objective.